app/uk/gov/hmrc/agentclientrelationships/controllers/DeAuthorisationRequestController.scala

```diff
@@ -0,0 +1,213 @@
+import uk.gov.hmrc.agentclientrelationships.connectors.{EnrolmentStoreProxyConnector, PirRelationshipConnector}
+import uk.gov.hmrc.agentclientrelationships.model.EnrolmentKey
+import uk.gov.hmrc.agentclientrelationships.model.invitation.InvitationFailureResponse.{ClientRegistrationNotFound, EnrolmentKeyNotFound, InvalidClientId, InvitationNotFound, RelationshipDeleteFailed, UnsupportedService}
+import uk.gov.hmrc.agentclientrelationships.model.invitation.{DeleteInvitationRequest, InvitationFailureResponse, ValidRequest}
+import uk.gov.hmrc.agentclientrelationships.services.{DeAuthorisationService, DeleteRelationshipsServiceWithAcr, RelationshipsCommon}
+import uk.gov.hmrc.agentmtdidentifiers.model.{Arn, Service}
+import uk.gov.hmrc.auth.core.AuthConnector
+import uk.gov.hmrc.http.{HeaderCarrier, UpstreamErrorResponse}
+import uk.gov.hmrc.play.bootstrap.backend.controller.BackendController
+
+import javax.inject.{Inject, Singleton}
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.control.NonFatal
+
+@Singleton
+class DeAuthorisationRequestController @Inject() (
+  deauthorisationService: DeAuthorisationService,
+  pirRelationshipConnector: PirRelationshipConnector,
+  deleteService: DeleteRelationshipsServiceWithAcr,
+  val esConnector: EnrolmentStoreProxyConnector,
+  val authConnector: AuthConnector,
+  val appConfig: AppConfig,
+  cc: ControllerComponents
+)(implicit ec: ExecutionContext)
+    extends BackendController(cc)
+    with AuthActions
+    with RelationshipsCommon {
+
+  val supportedServices: Seq[Service] = appConfig.supportedServices
+
+  def DeauthoriseRelationship(arn: Arn): Action[JsValue] = Action.async(parse.json) { implicit request =>
+    request.body
+      .validate[DeleteInvitationRequest]
+      .fold(
+        errs => Future.successful(BadRequest(s"Invalid payload: $errs")),
+        delInvReq => {
+          val responseT = for {
+
+            validRequest <-
+              EitherT.fromEither[Future](deauthorisationService.validateRequest(delInvReq.service, delInvReq.clientId))
+
+            result <- EitherT(deAuthInvitation(arn, validRequest))
+          } yield result
+
+          responseT.value
+            .map(
+              _.fold(
+                err => invitationErrorHandler(err, delInvReq.service, delInvReq.clientId),
+                result => result
+              )
+            )
+        }
+      )
+  }
+
+  private def deAuthInvitation(
+    arn: Arn,
+    validRequest: ValidRequest
+  )(implicit
+    hc: HeaderCarrier,
+    request: Request[Any]
+  ): Future[Either[InvitationFailureResponse, Result]] =
+    validRequest.service match {
+
+      case Service.PersonalIncomeRecord =>
+        pirRelationshipConnector
+          .deleteRelationship(arn, validRequest.service, validRequest.suppliedClientId.value)
+          .map {
+            case Some(true)  => Right(NoContent)
+            case Some(false) => Left(InvitationNotFound)
```

Comment:
also should this not be `RelationshipNotFound` as it reads like we would be failing deleteRelationship if no invitation was found when it's actually the relationship